libraries/crypto/src/hkdf.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::hmac::hmac_256;
+use super::{Hash256, HashBlockSize64Bytes};
+
+const HASH_SIZE: usize = 32;
+
+/// Computes the HKDF with empty salt and 256 bit (one block) output.
+pub fn hkdf_256<H>(ikm: &[u8], info: &[u8]) -> [u8; 32]
+where
+    H: Hash256 + HashBlockSize64Bytes,
+{
+    // Salt is a zero block here.
+    let prk = hmac_256::<H>(&[0; HASH_SIZE], ikm);
+    // l is implicitly 32, so we iterate exactly once.
+    let mut t = info.to_vec();
+    t.push(1);
```

Comment:
Note: in the future we may want to make the HMAC API more flexible (init/update/finish API) to allow for multiple updates and remove the need to allocate a Vec here.

Reply:
Noted. We have to reconsider the crypto API when we move away from software crypto eventually.